test/integration/dispatcher_integration_test.cc

```diff
@@ -177,13 +177,107 @@ TEST_P(DispatcherIntegrationTest, Basic) {
   ASSERT_EQ(cc.on_headers_calls, 1);
   ASSERT_EQ(cc.on_data_calls, 2);
   ASSERT_EQ(cc.on_complete_calls, 1);
+
+  test_server_->waitForCounterEq("http.dispatcher.stream_success", 1);
+}
+
+TEST_P(DispatcherIntegrationTest, BasicNon2xx) {
+  ConditionalInitializer ready_ran;
+  test_server_->server().dispatcher().post([this, &ready_ran]() -> void {
+    http_dispatcher_.ready(
+        test_server_->server().dispatcher(), test_server_->statStore(),
+        test_server_->server().listenerManager().apiListener()->get().http()->get());
+    ready_ran.setReady();
+  });
+  ready_ran.waitReady();
+
+  // Set response header status to be non-2xx to test correct stats get charged.
+  reinterpret_cast<AutonomousUpstream*>(fake_upstreams_.front().get())
+      ->setResponseHeaders(std::make_unique<Http::TestResponseHeaderMapImpl>(
+          Http::TestHeaderMapImpl({{":status", "503"}, {"content-length", "0"}})));
+
+  envoy_stream_t stream = 1;
+  // Setup bridge_callbacks to handle the response.
+  envoy_http_callbacks bridge_callbacks;
+  ConditionalInitializer terminal_callback;
+  callbacks_called cc = {0, 0, 0, 0, 0, &terminal_callback};
+  bridge_callbacks.context = &cc;
+  bridge_callbacks.on_headers = [](envoy_headers, bool, void* context) -> void {
+    callbacks_called* cc = static_cast<callbacks_called*>(context);
+    cc->on_headers_calls++;
+  };
+  bridge_callbacks.on_complete = [](void* context) -> void {
+    callbacks_called* cc = static_cast<callbacks_called*>(context);
+    cc->on_complete_calls++;
+    cc->terminal_callback->setReady();
+  };
+  bridge_callbacks.on_error = [](envoy_error, void* context) -> void {
+    callbacks_called* cc = static_cast<callbacks_called*>(context);
+    cc->on_error_calls++;
+  };
+
+  // Build a set of request headers.
+  Http::TestRequestHeaderMapImpl headers;
+  HttpTestUtility::addDefaultHeaders(headers);
+  envoy_headers c_headers = Http::Utility::toBridgeHeaders(headers);
+
+  // Create a stream.
+  EXPECT_EQ(http_dispatcher_.startStream(stream, bridge_callbacks), ENVOY_SUCCESS);
+  http_dispatcher_.sendHeaders(stream, c_headers, true);
+
+  terminal_callback.waitReady();
+
+  ASSERT_EQ(cc.on_error_calls, 0);
+  ASSERT_EQ(cc.on_headers_calls, 1);
+  ASSERT_EQ(cc.on_complete_calls, 1);
+
+  test_server_->waitForCounterEq("http.dispatcher.stream_failure", 1);
+}
+
+TEST_P(DispatcherIntegrationTest, BasicReset) {
```

Comment:
What do you think about the naming `BasicResetsIncrementFailures` to help describe what behavior we are trying to exercise here?